test/e2e/middleware-request-header-manipulation/test/index.test.ts

```diff
@@ -0,0 +1,102 @@
+/* eslint-env jest */
+
+import { join } from 'path'
+import { NextInstance } from 'test/lib/next-modes/base'
+import { fetchViaHTTP } from 'next-test-utils'
+import { createNext, FileRef } from 'e2e-utils'
+
+describe('Middleware Request Headers Manipulation', () => {
```

Comment:
Can you add a similar test for `test/e2e/app-dir`? The new router relies on certain headers to be available and this change might cause accidentally breaking the application if it doesn't copy them over, especially in the `Response.next(` case

Reply:
Added `app`-based tests `app-middleware`:

```
app-middleware
├── app
│   ├── layout.js
│   └── ssr-page.server.js
├── middleware.js
├── next.config.js
└── pages
    └── api
        ├── dump-headers-edge.js
        └── dump-headers-serverless.js
```

newbie question: Is gSSP available in the Layout RFC? Wonder if a SSR page can still read request headers. If not I'll just remove this test.

```js
export default function SSRPage({ headers }) {
  return (
    <>
      <p id="headers">{JSON.stringify(headers)}</p>
    </>
  )
}

export const getServerSideProps = (ctx) => {
  return {
    props: {
      headers: ctx.req.headers,
    },
  }
}
```

Reply:
Nvm I noticed `next/headers` in the latest canary. Will update the failing test 😄